Aleph Alpha's case study. (#2253)
mainmatter/mainmatter.com: src/cases/aleph-alpha.md

experience with these tools to Aleph Alpha's team, while still benefiting from the ease of use of a managed service. 
At the same time, it reduced the overall infrastructure risk for the project: if we were to run into an unsolvable
issue with StackIt's managed offering, we could always fall back to the Open Source version or switch to another
provider.

The final infrastructure stack looked like this:
- **Object storage**: we used StackIt's managed object storage to store the input and output data for the pipeline.
  The interface was compatible with the S3 API, which allowed us to rely on AWS' battle-tested Rust SDK.
- **Message broker**: we picked RabbitMQ to pass messages between the different components
  of the pipeline. [`lapin`](https://crates.io/crates/lapin) served well as a Rust client.
- **Metadata storage**: we used PostgreSQL to store metadata about the pipeline's progress. We relied on
  [`sqlx`](https://crates.io/crates/sqlx) to interact with the database from Rust.
- **Compute**: we relied on StackIt's managed Kubernetes offering to run the pipeline.

### Rust

> The entire data pipeline was built in Rust. That's the reason Aleph Alpha reached out to Mainmatter in the first place:

Reviewer: I'd mention Rust at the top where you talk about the need for performance?

Author: Where are you referring to? The architecture section?

Reviewer: The challenge I think

---
layout: case-study
company: Aleph Alpha
title: A large-scale data pipeline to feed AI models | Work
description:
  <p>Aleph Alpha empowers businesses and governments with the most advanced
  generative AI technology, to gain a decisive edge in the thriving AI
  economy.</p><p>Training AI models requires processing large amounts of data –
  ideally as efficiently as possible to save time and money. When Aleph Alpha
  set out to train a new model, they reached out to Mainmatter for support with
  architecting and implementing a capable and efficient data pipeline in
  Rust.</p>
permalink: false
---

## About Aleph Alpha

> Aleph Alpha is a German AI startup, a leader in the field of explainable and trustworthy sovereign AI.  

Reviewer 2: not sure what is meant by explainable? Maybe better word would be "accessible" or "understandable"

Author: This is straight from their press release/company brand content, I ported it verbatim 🤷🏻‍♂️

We took responsibility over four key areas: architecture, infrastructure, Rust and mentoring.

### Architecture

You need to **distribute the workload** across multiple machines to run such a large-scale pipeline to completion in a 
reasonable amount of time. That's how you fully leverage the capabilities of modern 
cloud computing: going from zero to **thousands of CPUs** for a few hours, then back to zero.  

The system as a whole must also satisfy a variety of other constraints:
- **Fault-tolerance**: if a machine fails, the pipeline should be able to recover and continue processing
  from where it left off, without losing any data.
- **Cost-efficiency**: storage fees, egress fees, compute fees... all of these can add up quickly.  
  The system must be designed with cost-efficiency in mind from the get-go.
- **Data lineage**: it should be possible to trace every piece of output data back to its source. This is
  critical for debugging and auditing purposes.

We worked closely with Aleph Alpha's team to design a system that would satisfy all of these constraints, while 
being **simple to operate** and **easy to reason about**.  

> 

Author: It's an engineering way of saying to refer to the quality/clarity of a chunk code/system—see https://softwareengineering.stackexchange.com/questions/351244/easy-to-reason-about-what-does-that-mean

---
layout: case-study
company: Aleph Alpha

> title: A large-scale data pipeline to feed AI models | Work

Author: A few alternatives I considered:
- "Preprocessing trillions of tokens with Rust"
- "Large models need large datasets: preprocessing trillions of tokens with Rust"

Reviewer: I think the first one is good – bit more catchy than the latter :)

Author: Yup, that's my favourite as well.

        plug the Python code into the Rust pipeline (thanks <a href="https://crates.io/crates/pyo3">pyo3</a>!) 
        to verify its functionality and run it at scale. If the change was desirable, and we needed better performance, 
        we would then port the Python code over to Rust. </li>
      </ul>
    </div>
</div>

<div class="case-study__section">
  <h2 class="case-study__heading h5">Mentoring</h2>
    <div class="case-study__text">
      <p>We split our time between delivery and mentoring.  
      Luca spent several hours a week pairing with Aleph Alpha's team, working together on the project, discussing 
      design decisions, reviewing code and sharing knowledge.</p>
      <p>Aleph Alpha's team is experienced and highly talented. Nonetheless, the entire project was on an extremely
      tight schedule: Having someone with the deep expertise on the team allowed them to <strong>move faster</strong> and <strong>avoid costly
      mistakes</strong>.</p>
    </div>
</div>

<div class="case-study__section">
  <h2 class="case-study__heading h5">Outcome</h2>
    <div class="case-study__text">

>         <p>We successfully assembled a high-quality multilingual dataset with more than <strong>4.5 trillion tokens</strong>. 

Author: ```suggestion
        <p>We successfully preprocessed <strong>4.5 trillion tokens</strong> to assemble a high-quality multilingual dataset. 
```
